### PR TITLE
[JENKINS-55582] Turn modules into plugins

### PR DESCRIPTION
**Pull request by kuisathaverat**
[JENKINS-55582](https://issues.jenkins-ci.org/browse/JENKINS-55582) Convert modules to plugins
[JENKINS-64107](https://issues.jenkins-ci.org/browse/JENKINS-64107) Convert SSHD module to a plugin
[JEP](https://github.com/jenkinsci/jep/pull/326) Convert modules to plugins

* This PR turns the ssh-module into a plugin
* ~It changes the groupId to `org.jenkins-ci.plugins`~
* Use BOM fro core dependencies
* Update git-changelist-maven-extension
* Bump the Core to the current weekly until decided on which version the change can integrate

### CODE REVIEW

**Comment by kuisathaverat** [comment]
Incrementals have failed I dunno why the artifacts are there and are correct 

https://ci.jenkins.io/blue/organizations/jenkins/Modules%2Fsshd-module/detail/PR-38/1/artifacts

Screenshot 2020-11-08 at 13 29 45


```
[2020-11-08T12:25:27.238Z] Invalid archive retrieved from Jenkins, perhaps the plugin is not properly incrementalized?

[2020-11-08T12:25:27.338Z] Error: ZIP error: Error: No permissions for org/jenkins-ci/plugins/sshd/3.0-rc161.0462582585e6/sshd-3.0-rc161.0462582585e6-javadoc.jar from https://ci.jenkins.io/job/Modules/job/sshd-module/job/PR-38/1/artifact/**/*.0462582585e6/*.0462582585e6*/*zip*/archive.zip
```

**Comment by kuisathaverat** [comment]
I think the incremental fail because of the change of the groupId.

File: pom.xml
Comment (jglick) [inline pom.xml:11]: Revert this. That is why you could not publish incrementals.

Comment (kuisathaverat) [reply pom.xml:11]: yep, I know, but we want to change the groupdId, isn't it?

Comment (jglick) [reply pom.xml]: The JEP proposes leaving it alone. There is no requirement that a plugin have a specific kind of `groupId`. See `maven-plugin` for example.

**Comment by kuisathaverat** [comment]
The PR passes the CI on Linux with JDK 11 and windows with JDK 8, on Linus with JDK8 the build is a success but there was some environmental issue with the agent so I think that is ready to merge.

```
[2020-11-22T19:55:03.950Z] [INFO] ------------------------------------------------------------------------
[2020-11-22T19:55:03.950Z] [INFO] BUILD SUCCESS
[2020-11-22T19:55:03.950Z] [INFO] ------------------------------------------------------------------------
[2020-11-22T19:55:03.950Z] [INFO] Total time:  04:20 min
[2020-11-22T19:55:03.950Z] [INFO] Finished at: 2020-11-22T19:55:03Z
[2020-11-22T19:55:03.950Z] [INFO] ------------------------------------------------------------------------
[2020-11-22T19:55:04.363Z] [Pipeline] }
[2020-11-22T19:55:04.390Z] [Pipeline] // withEnv
[2020-11-22T19:55:04.462Z] [Pipeline] junit
[2020-11-22T19:55:04.486Z] Recording test results
[2020-11-22T19:55:08.374Z] [GitHub Checks] GitHub check (name: Tests, status: COMPLETED) has been published.
[2020-11-22T19:55:08.374Z] [Pipeline] }
[2020-11-22T19:55:08.423Z] [Pipeline] // stage
[2020-11-22T19:55:08.451Z] [Pipeline] stage
[2020-11-22T19:55:08.452Z] [Pipeline] { (Archive (windows-8-2.164.1))
[2020-11-22T19:55:08.474Z] [Pipeline] echo
[2020-11-22T19:55:08.474Z] Skipping static analysis results for windows-8-2.164.1
[2020-11-22T19:55:08.474Z] [Pipeline] }
[2020-11-22T19:55:08.514Z] [Pipeline] // stage
[2020-11-22T19:55:08.536Z] [Pipeline] }
[2020-11-22T19:55:08.558Z] [Pipeline] // timeout
[2020-11-22T19:55:08.592Z] [Pipeline] }
[2020-11-22T19:55:08.611Z] [Pipeline] // node
[2020-11-22T19:55:08.632Z] [Pipeline] }
[2020-11-22T20:02:34.792Z] EC2 (aws) - High memory ubuntu 18.04  (i-041fee315b6b50f49) was marked offline: Connection was broken: java.io.EOFException
[2020-11-22T20:02:34.792Z] 	at java.base/java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2842)
[2020-11-22T20:02:34.792Z] 	at java.base/java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3337)
[2020-11-22T20:02:34.792Z] 	at java.base/java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:925)
[2020-11-22T20:02:34.792Z] 	at java.base/java.io.ObjectInputStream.<init>(ObjectInputStream.java:368)
[2020-11-22T20:02:34.792Z] 	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:49)
[2020-11-22T20:02:34.792Z] 	at hudson.remoting.Command.readFrom(Command.java:142)
[2020-11-22T20:02:34.792Z] 	at hudson.remoting.Command.readFrom(Command.java:128)
[2020-11-22T20:02:34.792Z] 	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)
[2020-11-22T20:02:34.792Z] 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)
[2020-11-22T20:02:34.792Z] Caused: java.io.IOException: Unexpected termination of the channel
[2020-11-22T20:02:34.792Z] 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:77)
[2020-11-22T20:02:34.792Z] 
[2020-11-22T20:02:34.793Z] [Pipeline] // node
[2020-11-22T20:02:34.834Z] [Pipeline] }
[2020-11-22T20:02:34.835Z] Failed in branch linux-8
[2020-11-22T20:02:34.862Z] [Pipeline] // parallel
[2020-11-22T20:02:34.884Z] [Pipeline] End of Pipeline
[2020-11-22T20:02:35.266Z] [GitHub Checks] GitHub check (name: Jenkins, status: COMPLETED) has been published.
[2020-11-22T20:02:35.988Z] 
[2020-11-22T20:02:35.988Z] GitHub has been notified of this commit’s build result
[2020-11-22T20:02:35.988Z] 
[2020-11-22T20:02:36.221Z] java.io.EOFException
[2020-11-22T20:02:36.221Z] 	at java.base/java.io.ObjectInputStream$PeekInputStream.readFully(ObjectInputStream.java:2842)
[2020-11-22T20:02:36.221Z] 	at java.base/java.io.ObjectInputStream$BlockDataInputStream.readShort(ObjectInputStream.java:3337)
[2020-11-22T20:02:36.221Z] 	at java.base/java.io.ObjectInputStream.readStreamHeader(ObjectInputStream.java:925)
[2020-11-22T20:02:36.221Z] 	at java.base/java.io.ObjectInputStream.<init>(ObjectInputStream.java:368)
[2020-11-22T20:02:36.221Z] 	at hudson.remoting.ObjectInputStreamEx.<init>(ObjectInputStreamEx.java:49)
[2020-11-22T20:02:36.221Z] 	at hudson.remoting.Command.readFrom(Command.java:142)
[2020-11-22T20:02:36.221Z] 	at hudson.remoting.Command.readFrom(Command.java:128)
[2020-11-22T20:02:36.221Z] 	at hudson.remoting.AbstractSynchronousByteArrayCommandTransport.read(AbstractSynchronousByteArrayCommandTransport.java:35)
[2020-11-22T20:02:36.221Z] 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:63)
[2020-11-22T20:02:36.221Z] Caused: java.io.IOException: Unexpected termination of the channel
[2020-11-22T20:02:36.221Z] 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:77)
[2020-11-22T20:02:36.221Z] Also:   hudson.remoting.Channel$CallSiteStackTrace: Remote call to EC2 (aws) - High memory ubuntu 18.04  (i-041fee315b6b50f49)
[2020-11-22T20:02:36.221Z] 		at hudson.remoting.Channel.attachCallSiteStackTrace(Channel.java:1800)
[2020-11-22T20:02:36.221Z] 		at hudson.remoting.Request.call(Request.java:198)
[2020-11-22T20:02:36.221Z] 		at hudson.remoting.Channel.call(Channel.java:1000)
[2020-11-22T20:02:36.221Z] 		at hudson.EnvVars.getRemote(EnvVars.java:437)
[2020-11-22T20:02:36.221Z] 		at hudson.model.Computer.getEnvironment(Computer.java:1193)
[2020-11-22T20:02:36.221Z] 		at org.jenkinsci.plugins.workflow.support.steps.ExecutorStepExecution$PlaceholderTask$PlaceholderExecutable.run(ExecutorStepExecution.java:801)
[2020-11-22T20:02:36.221Z] 		at hudson.model.ResourceController.execute(ResourceController.java:97)
[2020-11-22T20:02:36.221Z] 		at hudson.model.Executor.run(Executor.java:428)
[2020-11-22T20:02:36.221Z] Also:   hudson.model.Computer$TerminationRequest: Termination requested at Sun Nov 22 19:50:47 UTC 2020 by Thread[Ping thread for channel hudson.remoting.Channel@5f0bcef:EC2 (aws) - High memory ubuntu 18.04  (i-041fee315b6b50f49),5,main] [id=1005376]
[2020-11-22T20:02:36.221Z] 		at hudson.model.Computer.recordTermination(Computer.java:229)
[2020-11-22T20:02:36.221Z] 		at hudson.model.Computer.disconnect(Computer.java:495)
[2020-11-22T20:02:36.221Z] 		at hudson.slaves.SlaveComputer.disconnect(SlaveComputer.java:759)
[2020-11-22T20:02:36.221Z] 		at hudson.slaves.ChannelPinger$1.onDead(ChannelPinger.java:198)
[2020-11-22T20:02:36.221Z] 		at hudson.remoting.PingThread.ping(PingThread.java:134)
[2020-11-22T20:02:36.221Z] 		at hudson.remoting.PingThread.run(PingThread.java:90)
[2020-11-22T20:02:36.221Z] Caused: hudson.remoting.RequestAbortedException
[2020-11-22T20:02:36.221Z] 	at hudson.remoting.Request.abort(Request.java:344)
[2020-11-22T20:02:36.221Z] 	at hudson.remoting.Channel.terminate(Channel.java:1085)
[2020-11-22T20:02:36.221Z] 	at hudson.remoting.SynchronousCommandTransport$ReaderThread.run(SynchronousCommandTransport.java:94)
[2020-11-22T20:02:36.221Z] Finished: FAILURE
```

**Review by jglick — APPROVED** [review]
Looks good from a code review POV. My main request is that the full change be sanity-checked carefully, as outlined in the JEP and https://github.com/jenkinsci/jenkins/pull/3988, using `java -war jenkins.war` (_not_ `JenkinsRule`):

----

SSH functionality was also tested. With the new `sshd` plugin enabled, you can configure a port, attach your public key to your user configuration, and run commands like

```bash
ssh -p 2222 admin@localhost who-am-i
```

You can also install `workflow-cps-global-lib` and use the (now deprecated) in-master Git server, `workflowLibs.git`, via SSH protocol.

**Comment by kuisathaverat** [comment]
> Looks good from a code review POV. My main request is that the full change be sanity-checked carefully, as outlined in the JEP and [jenkinsci/jenkins#3988](https://github.com/jenkinsci/jenkins/pull/3988), using `java -war jenkins.war` (_not_ `JenkinsRule`):
> 
> SSH functionality was also tested. With the new `sshd` plugin enabled, you can configure a port, attach your public key to your user configuration, and run commands like
> 
> ```shell
> ssh -p 2222 admin@localhost who-am-i
> ```
> 
> You can also install `workflow-cps-global-lib` and use the (now deprecated) in-master Git server, `workflowLibs.git`, via SSH protocol.

@jglick 
I've configured a test environment and make some manual tests, Where is the best place to put those tests (or a kind of) to run them automatically?

```
git clone https://github.com/kuisathaverat/jenkins-issues.git
cd jenkins-issues/sshd-module/
make download-dependencies
make start
make test
```

**Comment by jglick** [comment]
> run them automatically

That would be a major endeavor. I do not think the acceptance test harness is set up for this sort of client interaction. My request was just that somebody, at some point, tested all this in a realistic context and ensured that everything behaves like it did.

**Comment by kuisathaverat** [comment]
I will try to convert it into something launched from JUnit, then I will add those 3 tests to the sshd-module.

**Comment by jglick** [comment]
> those 3 tests

Sorry, which three tests?

**Comment by kuisathaverat** [comment]
> > those 3 tests
> 
> Sorry, which three tests?

these tests https://github.com/kuisathaverat/jenkins-issues/blob/master/sshd-module/Makefile#L69-L79

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
ready for this to be shipped?

**Comment by kuisathaverat** [comment]
> ready for this to be shipped?

sure thing

**Comment by timja** [comment]
@daniel-beck @jvz @oleg-nenashev you seem to be listed for release here.

Do one of you want to ship it? Or I can add myself if preferred

**Review by jvz — APPROVED** [review]

**Comment by timja** [comment]
@jvz do you plan to ship a release with this change?

**Comment by jvz** [comment]
I haven't made a release of a Jenkins module before, so I'll defer to you and @daniel-beck here.

File: pom.xml
Comment (jglick) [inline pom.xml:34]: Update this before merge.

File: pom.xml
Comment (jglick) [inline pom.xml:42]: Pick an appropriate line.

File: src/main/java/org/jenkinsci/main/modules/cli/auth/ssh/KeyEncodeHelper.java
Comment (jglick) [inline src/main/java/org/jenkinsci/main/modules/cli/auth/ssh/KeyEncodeHelper.java:20]: Do not forget to archive https://github.com/jenkinsci/ssh-cli-auth-module

File: pom.xml
Comment (timja) [reply pom.xml:34]:
```suggestion
    <jenkins.version>2.280</jenkins.version> <!-- Current weekly -->
```

File: pom.xml
Comment (timja) [reply pom.xml:42]:
```suggestion
        <artifactId>bom-2.277.x</artifactId>
```

File: pom.xml
Comment (timja) [inline pom.xml:43]:
```suggestion
        <version>25</version>
```

**Comment by timja** [comment]
https://github.com/jenkinsci/sshd-module/releases/tag/sshd-3.0

**Comment by jglick** [comment]
I suppose this repo should also be renamed `sshd-plugin` and https://github.com/jenkinsci/sshd-module/blob/637aa7ef19825fbbd240b1c3a46aff107d3a3321/pom.xml#L18-L20 updated accordingly?

**Comment by timja** [comment]
done

**Pull request merged**
